[API Integration] Walters Art Museum Provider API Script and Test Suite

dravadhis opened the pull request:
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, open one before creating this pull request. -->
Fixes #274 by @annatuma 

## Description
<!-- Concisely describe what the pull request does. -->
This PR adds provider API script and the test suite for the script for integrating the Walters Art Museum API.


## Technical details
<!-- Add any other information or technical details about the implementation; or delete the section entirely. -->
Apart from the script and the test suite, following files have been added/edited:
1. [Added] src/cc_catalog_airflow/dags/provider_api_scripts/test/resources/waltersartmuseum: Test resources for test suite of the script
2. [Edited] src/cc_catalog_airflow/dags/util/loader/provider_details.py: Added provider Walters Art Museum

## Tests
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete the section entirely. -->

## Screenshots
<!-- Add screenshots to show the problem and the solution; or delete the section entirely. -->

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the *default* branch of the repository (`main` or `master`).
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

Commit:
Integrate Walters Art Museum API with CC Search

Commit:
Integrate Walters Art Museum API

kss682 commented on src/cc_catalog_airflow/env.template:
```diff
@@ -1,6 +1,7 @@
 # Change OUTPUT_DIR as appropriate for production
 OUTPUT_DIR=/tmp/
 
+WALTERS_ART_MUSEUEM_KEY=SXmSvc92Lx6hngcJbVGEWPHVz1wKAxboloKJqHMjky4S9POeQMSVAsg3GHSugorN
```
We are not supposed upload the token in the template, set it as not_set.

Commit:
Integrate Walters Art Museum API with CC Search
Set API KEY in env.template as not_set

mathemancer commented on src/cc_catalog_airflow/dags/provider_api_scripts/walters_art_museum.py:
```diff
@@ -0,0 +1,268 @@
+    )
+
+
+def _get_foreign_landing_url(img):
+    foreign_landing_url = None
+    foreign_landing_url = img.get("ResourceURL")
+    return foreign_landing_url
```
This function has precisely the same effect as `img.get("ResourceURL")`.  This is because `None` is the default return value for the `dict.get` method of python.  Thus, all of these 'get' functions should be removed, and replaced with the bare call to `dict.get` where each is called.

mathemancer commented on src/cc_catalog_airflow/dags/provider_api_scripts/walters_art_museum.py:
```diff
@@ -0,0 +1,268 @@
+    creator, creator_url = None, None
+    creator = img.get("Creator")
+    if creator:
+        creator = img.get("Creator")
```
This is redundant, since `creator` has already been set to the required value

mathemancer commented on src/cc_catalog_airflow/dags/provider_api_scripts/walters_art_museum.py:
```diff
@@ -0,0 +1,268 @@
+    logger.debug(f'Processing Image: {img}')
+
+    foreign_landing_url = _get_foreign_landing_url(img)
+    image_url = _get_image_url(img)
```
```suggestion
    image_url = img.get("PrimaryImage", {}).get("Raw")
```
Please use this form for any 'Double get' calls (to replace the wrapper functions).  Note that this keeps things from failing in the event `"PrimaryImage"` isn't a key in the given `img` dict (see the replacement of the default `None` return with a default `{}` return)

mathemancer commented on src/cc_catalog_airflow/dags/provider_api_scripts/walters_art_museum.py:
```diff
@@ -0,0 +1,268 @@
+    image_meta_data["Classification"] = img.get("Classification")
+    image_meta_data["Description"] = img.get("Description")
+    image_meta_data["CreditLine"] = img.get("CreditLine")
+    return {k: v for k, v in image_meta_data.items() if v is not None}
```
Nice!

mathemancer commented on src/cc_catalog_airflow/dags/provider_api_scripts/walters_art_museum.py:
```diff
@@ -0,0 +1,268 @@
+    logger.info("Terminating Script!")
+
+
+def _get_image_list(
```
I'd like for you to make an attempt at refactoring this function with the following 2 goals:
1. Please try to use the higher-level `get_response_json` method of the `DelayedRequester` class.  It does the checks you're performing here out-of-the-box.
2. Please leave the retrying logic to that function.  This will avoid the nested loops here, and make the overall flow easier to follow.

mathemancer reviewed (changes requested):
Overall, pretty good.  Thorough tests; thanks for that!

My major requests are to 
1. Please remove the redundant wrappers around `dict.get` (note that not all are redundant, but most are).
2. Please refactor the `_get_image_list` function as noted in the specific comment.

Thank you!

Commit:
Integrate Walters Art Museum API with CC Search
Modify _get_image_list to work with _get_response_json method
of the DelayedRequester class. Replace get functions with bare
dict.get calls.

mathemancer commented on src/cc_catalog_airflow/dags/provider_api_scripts/walters_art_museum.py:
```diff
@@ -0,0 +1,209 @@
+LIMIT = 250000
+PROVIDER = prov.WALTERS_DEFAULT_PROVIDER
+REQUEST_TYPE = 'objects'
+ENDPOINT = f'http://api.thewalters.org/v1/{REQUEST_TYPE}'
```
Please use SSL/TLS when possible.
```suggestion
ENDPOINT = f'https://api.thewalters.org/v1/{REQUEST_TYPE}'

```

mathemancer commented on src/cc_catalog_airflow/dags/provider_api_scripts/walters_art_museum.py:
```diff
@@ -0,0 +1,209 @@
+
+logging.basicConfig(
+    format='%(asctime)s - %(name)s - %(levelname)s:  %(message)s',
+    level=logging.DEBUG
```
Please turn down the logging level.

```suggestion
    level=logging.INFO

```

mathemancer commented on src/cc_catalog_airflow/dags/provider_api_scripts/walters_art_museum.py:
```diff
@@ -0,0 +1,209 @@
+        for img in items_list:
+            image_list.append(img)
+
+        if json_response_inpydict_form.get("NextPage") is False:
```
This will loop indefinitely in the case where "NextPage" is missing from the JSON for any reason.  Please use a more robust check (it's better to break unnecessarily than send indefinite identical requests to a 3rd party API).
```suggestion
        if json_response_inpydict_form.get("NextPage") is not True:

```

mathemancer reviewed (changes requested):
This is very close to complete!

Just a couple of details I noticed that need to change.

Thank you!

mathemancer commented:
@dravadhis Ah, I just noticed a major problem here.  You need to increment the page you're requesting.  In the current logic, the main while loop just requests the first page over and over.  Please fix that as well before re-requesting review!

Thanks.

Commit:
Integrate Walters Art Museum API with CC Search
Implement page increment logic in walters_art_museum.py

Commit:
Integrate Walters Art Museum API with CC Search

Commit:
Integrate Walters Art Museum API with CC Search

mathemancer reviewed (changes requested):
Looking great!

The only problem is that the classification `'Paper & Paper-Mache'` seems to break things.  Please either figure out why that's so and fix it, or just remove that classification for now.  

Thank you!

mathemancer commented:
@dravadhis You can see the breakage by just running the script with *only* the defective classification in the list.

Commit:
Integrate Walters Art Museum API with CC Search
Remove 'Paper & Paper-Mache' from list of classifications.

dravadhis commented:
> Looking great!
> 
> The only problem is that the classification `'Paper & Paper-Mache'` seems to break things. Please either figure out why that's so and fix it, or just remove that classification for now.
> 
> Thank you!

I checked the responses. Turns out the classification has been removed but not updated in the documentation yet. I have removed it from the classification list (`QUERY_CLASSIFICATION`) list in the`walters_art_museum.py` script.

mathemancer reviewed (approved):
Looks great!

Thank you for all your work on this.
